### PR TITLE
tracer: enforce trace policy entirely via Sampler

### PR DESCRIPTION
**Pull request by bobheadxi**
This change updates our internal tracer to enforce policy only via a `Sampler` implementation. This has the following benefits:

1. Even when a trace should not be sampled, contexts are still populated with valid spans, rather than no-op ones. This is important to make use of trace IDs for non-tracing purposes, e.g. https://github.com/sourcegraph/sourcegraph/pull/57774 and https://github.com/sourcegraph/sourcegraph/pull/58060
2. We enforce trace policies the way they were meant to be enforced in OpenTelemetry: by simply indicating that the span should not be exported.

This was not possible before because OpenTracing did not use context propagation, hence we did not have a way to use trace policy flags set in context - but thanks to @camdencheek's work removing OpenTracing entirely, we can now do this in a more idiomatic fashion.

Thanks to this, I've removed a few places that prevented trace context from being populated based on trace policy (HTTP and GraphQL middleware, and `internal/trace`). This delegates sampling decisions to the sampler, and ensures we accept valid trace context everywhere.

## Test plan

Unit tests on a TracerProvider configured with the new sampler.

Manual testing:

```
sg run jaeger otel-collector 
sg start
```

Setting `observability.tracing.debug` to `true` we can see logs indicating the desired traits for non-`ShouldTrace` traces:

```
[         worker] INFO tracer tracer/logged_otel.go:63 Start {"spanName": "workerutil.dbworker.store.insights_query_runner_jobs_store.dequeue", "isRecording": false, "isSampled": false, "isValid": true}
```

With `observability.tracing.sampling` set to `none`, running a search with `&trace=1` only gives us spans from zoekt, which seems to have always been outside our rules here.

With `observability.tracing.sampling` set to `selective`, running a search with `&trace=1` gives us a full trace.

With `observability.tracing.sampling` set to `all`, Jaeger instantly gets loads of traces, and in logs we see:

```
[         worker] INFO tracer tracer/logged_otel.go:63 Start {"spanName": "workerutil.dbworker.store.exhaustive_search_worker_store.dequeue", "isRecording": true, "isSampled": true, "isValid": true}
```

### CODE REVIEW

File: internal/trace/policy/policy.go [hunk 1]
```diff
@@ -38,9 +38,15 @@ type key int
 
 const shouldTraceKey key = iota
 
-// ShouldTrace returns true if the shouldTraceKey context value is true. It is used to
-// determine if a trace should be started by various middleware. If the value is not set
-// at all, we check if we should the global policy is set to TraceAll instead.
+// ShouldTrace returns true if the shouldTraceKey context value is true. If the
+// value is not set at all, we check if we should the global policy is set to
```
Comment (burmudar) [inline internal/trace/policy/policy.go:42]:
```suggestion
// value is not set at all, we check if what the global policy is set to
```

Comment (burmudar) [inline internal/trace/policy/policy.go:42]:
```suggestion
// value is not set at all, we check if the global policy is set to
```

**Review by burmudar — APPROVED** [review]
Nice!

File: cmd/frontend/graphqlbackend/graphqlbackend.go [hunk 2]
```diff
@@ -182,10 +180,7 @@ func (requestTracer) TraceField(ctx context.Context, _, typeName, fieldName stri
 }
 
 func (t requestTracer) TraceValidation(ctx context.Context) func([]*gqlerrors.QueryError) {
-	var finish func([]*gqlerrors.QueryError)
-	if policy.ShouldTrace(ctx) {
-		finish = t.tracer.TraceValidation(ctx)
-	}
+	finish := t.tracer.TraceValidation(ctx)
 	return func(queryErrors []*gqlerrors.QueryError) {
 		if finish != nil {
```
Comment (camdencheek) [inline cmd/frontend/graphqlbackend/graphqlbackend.go:185]: In all these spots, `finish` will always be non-nil now, right?

**Review by camdencheek — APPROVED** [review]
very exciting!

File: cmd/frontend/graphqlbackend/graphqlbackend.go (hunk 2 above)
Comment (bobheadxi) [reply cmd/frontend/graphqlbackend/graphqlbackend.go]: You're right, simplified it!

**Comment by bobheadxi** [comment]
Backporting as bugfix for trace ID propagation

**Comment by bobheadxi** [comment]
Manual backport: https://github.com/sourcegraph/sourcegraph/pull/58084

(logger diffs 😰 )

**Pull request merged**